src/datasources/push-notifications-api/firebase-cloud-messaging-api.service.ts

```diff
@@ -0,0 +1,141 @@
+    'https://www.googleapis.com/auth/firebase.messaging';
+
+  private readonly baseUrl: string;
+  private readonly project: string;
+  private readonly clientEmail: string;
+  private readonly privateKey: string;
+
+  constructor(
+    @Inject(NetworkService)
+    private readonly networkService: INetworkService,
+    @Inject(IConfigurationService)
+    private readonly configurationService: IConfigurationService,
+    @Inject(CacheService)
+    private readonly cacheService: ICacheService,
+    private readonly httpErrorFactory: HttpErrorFactory,
+  ) {
+    this.baseUrl = this.configurationService.getOrThrow<string>(
+      'pushNotifications.baseUri',
+    );
+    this.project = this.configurationService.getOrThrow<string>(
+      'pushNotifications.project',
+    );
+    // Service account credentials are used for OAuth2 assertion
+    this.clientEmail = this.configurationService.getOrThrow<string>(
+      'pushNotifications.serviceAccount.clientEmail',
+    );
+    this.privateKey = this.configurationService.getOrThrow<string>(
+      'pushNotifications.serviceAccount.privateKey',
+    );
+  }
+
+  /**
+   * Enqueues a notification to be sent to a device with given FCM token.
+   *
+   * @param fcmToken - device's FCM token
+   * @param notification - notification payload
+   */
+  async enqueueNotification<T extends Record<string, unknown>>(
+    fcmToken: string,
+    notification: FirebaseNotification<T>,
+  ): Promise<void> {
+    const url = `${this.baseUrl}/${this.project}/messages:send`;
+    try {
+      const accessToken = await this.getOauth2Token();
+      await this.networkService.post({
+        url,
+        data: {
+          message: {
+            token: fcmToken,
+            notification,
+          },
+        },
+        networkRequest: {
+          headers: {
+            Authorization: `Bearer ${accessToken}`,
+          },
+        },
+      });
+    } catch (error) {
+      /**
+       * TODO: Error handling based on `error.details[i].reason`, e.g.
+       * - expired OAuth2 token
+       * - stale FCM token
+       */
+      throw this.httpErrorFactory.from(error);
+    }
+  }
+
+  /**
+   * Retrieves and caches OAuth2 token for Firebase Cloud Messaging API.
+   *
+   * @returns - OAuth2 token
+   */
+  private async getOauth2Token(): Promise<string> {
+    const cacheDir = CacheRouter.getFirebaseOAuth2TokenCacheDir();
+    const cachedToken = await this.cacheService.get(cacheDir);
+
+    if (cachedToken) {
+      return cachedToken;
+    }
+
+    const { data } = await this.networkService.post<{
+      access_token: string;
+      expires_in: number;
+      token_type: string;
+    }>({
+      url: FirebaseCloudMessagingApiService.OAuth2TokenUrl,
+      data: {
+        grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
+        assertion: this.getAssertion(),
+      },
+    });
+
+    // Token cached according to issuance
+    await this.cacheService.set(cacheDir, data.access_token, data.expires_in);
```

Comment:
Very nitpicky: should we store `expires_in - (some seconds)` to cover the edge case where for some reason we spend some seconds on storing the token after it was generated (network delays, etc.)

Reply:
Added in 8afe78e433a284ed1500d09d16cb71fe2e6da725.